File: routeMap/app/src/main/java/ru/android/zheka/route/BellmannFord.java
package ru.android.zheka.route;

import com.google.android.gms.maps.model.LatLng;

import java.util.ArrayList;
import java.util.Arrays;

import ru.android.zheka.gmapexample1.MapsActivity;

public class BellmannFord {
    final static double toRad = Math.PI/180;
    final static double R = 6371;//km
    // !!! use process once
    public static double length = 0;
    public static float bearing=-1000;
    static double longitude1;
    static double longitude;
    static double latitude1;
    static double latitude;
    static double dr;
    static double dr_tay;
    static private double eps = 1.0E3;

    public static LatLng[] process(LatLng[] latLngs){
        LatLng [] order = fliplr (latLngs);
        order = getOrder(order);
        order = fliplr (order);
        order = getOrder(order);
        length = getPathLength (order);
        System.out.println ("BellmanFord opt length = "+length);
        return order;
    }

    private static LatLng[] getOrder(LatLng[] latLngs) {
        int n = latLngs.length;
        LatLng [] reOrder = latLngs;
        boolean whileTrue = true;
        double optLen = getPathLength(reOrder);
        while (whileTrue){
            whileTrue = false;
            for (int i = 1; i < n-1; i++)// do not touch end point
                for (int j=i+1; j < n-1; j++){
                    LatLng tmp = reOrder[j];
                    reOrder[j] = reOrder[i];
                    reOrder[i] = tmp;
                    double curLen = getPathLength(reOrder);
                    if (curLen < optLen){
                        optLen = curLen;
                        whileTrue = true;
                        System.out.println ("new opt: optLen="+optLen);
                    }else{
                        tmp = reOrder[j];
                        reOrder[j] = reOrder[i];
                        reOrder[i] = tmp;
                    }
            }
        }
        return reOrder;
    }

    public static double getPathLength(LatLng[] latLngs){
        int n = latLngs.length;
        double pathLen = 0;
        if (n==0)
            return pathLen;
        double longitude = latLngs[0].longitude * toRad;
        double latitude = latLngs[0].latitude * toRad;
        for (int j = 1; j < n; j++) {
            double latitude1 = latLngs[j].latitude * toRad;
            double longitude1 = latLngs[j].longitude * toRad;
            double dLon = -(longitude1 - longitude);
            double dLamda = latitude1 - latitude;
            double cosArg = Math.cos (dLamda) + (Math.cos (dLon) - 1.0)
                    * Math.cos (latitude) * Math.cos (latitude1);
            latitude = latitude1;
            longitude = longitude1;
            double arg = Math.sqrt (1 - cosArg * cosArg) / cosArg;
            // atg Tailor for small arg
            arg = arg - 1.0 / 3 * arg * arg * arg;
            pathLen += R * Math.abs (arg);
        }
        return pathLen;
    }
    static private LatLng[] fliplr(LatLng[] points){
        int n = points.length;
        LatLng[] inv = new LatLng [n];
        for (int i = 0;i < n;i++){
            inv[n-i-1] = points[i];
        }
        return inv;
    }
    static public float getBearing(LatLng[] points) throws NoDirectionException, MissMatchDataException {
        if (points.length<2)
            throw new MissMatchDataException ();
        LatLng point = round (points[0]);
        LatLng point1 = round (points[1]);
        if (point.equals (point1)){
            int i=2;
            while(point.equals (point1) && i<points.length){
                point1 = round (points[i++]);
            }
            if (point.equals (point1))
                throw  new NoDirectionException ();
        }
        longitude1 = point1.longitude*toRad;
        longitude = point.longitude*toRad;
        latitude1 = point1.latitude*toRad;
        latitude = point.latitude*toRad;
        double dLon = longitude1 - longitude;
        double dLambda = latitude1 - latitude;
        double dr2 = 2.0*(1 + (1.0 - Math.cos (dLon))*Math.cos(latitude1)*Math.cos (latitude)
            - Math.cos (dLambda));
        dr = Math.sqrt (dr2);
        dr_tay = Math.sin(dLambda) - (1 - Math.cos (dLon))*Math.cos (latitude1)*Math.cos (latitude);
        //if (dr==0)
        //    throw  new NoDirectionException ();
        bearing = (float) (Math.acos (dr_tay/dr)/toRad); // 2 roots
        if (dLon<0)
            bearing = (float) 360.0 - bearing;
        System.out.println ("bearing="+bearing+" points="+ Arrays.asList (points));
        return bearing;
    }
    public static LatLng round(LatLng point){
        double lat = Math.round (point.latitude*eps)/eps;
        double lon = Math.round (point.longitude*eps)/eps;
        return new LatLng (lat,lon);
    }
    static public class MissMatchDataException extends Exception{

    }

    static public class NoDirectionException extends Exception{

    }
}

File: routeMap/app/src/main/java/ru/android/zheka/route/BellmannFord.kt
package ru.android.zheka.route

import com.google.android.gms.maps.model.LatLng
import java.util.*

object BellmannFord {
    const val toRad = Math.PI / 180
    const val R = 6371.0 //km

    // !!! use process once
    @JvmField
    var length = 0.0
    @JvmField
    var bearing = -1000f
    var longitude1 = 0.0
    var longitude = 0.0
    var latitude1 = 0.0
    var latitude = 0.0
    var dr = 0.0
    var dr_tay = 0.0
    private const val eps = 1.0E3
    @JvmStatic
    fun process(latLngs: Array<LatLng>): Array<LatLng> {
        var order = fliplr(latLngs)
        order = getOrder(order)
        order = fliplr(order)
        order = getOrder(order)
        length = getPathLength(order)
        println("BellmanFord opt length = $length")
        return order
    }

    private fun getOrder(latLngs: Array<LatLng>): Array<LatLng> {
        val n = latLngs.size
        var whileTrue = true
        var optLen = getPathLength(latLngs)
        while (whileTrue) {
            whileTrue = false
            for (i in 1 until n - 1)  // do not touch end point
                for (j in i + 1 until n - 1) {
                    var tmp = latLngs[j]
                    latLngs[j] = latLngs[i]
                    latLngs[i] = tmp
                    val curLen = getPathLength(latLngs)
                    if (curLen < optLen) {
                        optLen = curLen
                        whileTrue = true
                        println("new opt: optLen=$optLen")
                    } else {
                        tmp = latLngs[j]
                        latLngs[j] = latLngs[i]
                        latLngs[i] = tmp
                    }
                }
        }
        return latLngs
    }

    fun getPathLength(latLngs: Array<out LatLng>): Double {
        val n = latLngs.size
        var pathLen = 0.0
        if (n == 0) return pathLen
        var longitude = latLngs[0].longitude * toRad
        var latitude = latLngs[0].latitude * toRad
        for (j in 1 until n) {
            val latitude1 = latLngs[j].latitude * toRad
            val longitude1 = latLngs[j].longitude * toRad
            val dLon = -(longitude1 - longitude)
            val dLamda = latitude1 - latitude
            val cosArg = Math.cos(dLamda) + ((Math.cos(dLon) - 1.0)
                    * Math.cos(latitude) * Math.cos(latitude1))
            latitude = latitude1
            longitude = longitude1
            var arg = Math.sqrt(1 - cosArg * cosArg) / cosArg
            // atg Tailor for small arg
            arg = arg - 1.0 / 3 * arg * arg * arg
            pathLen += R * Math.abs(arg)
        }
        return pathLen
    }

    private fun fliplr(points: Array<LatLng>): Array<LatLng> {
        val n = points.size
        val inv = Array(n){i-> points[n - i - 1]}
        return inv
    }

    @Throws(NoDirectionException::class, MissMatchDataException::class)
    fun getBearing(points: Array<LatLng>): Float {
        if (points.size < 2) throw MissMatchDataException()
        val point = round(points[0])
        var point1 = round(points[1])
        if (point == point1) {
            var i = 2
            while (point == point1 && i < points.size) {
                point1 = round(points[i++])
            }
            if (point == point1) throw NoDirectionException()
        }
        longitude1 = point1.longitude * toRad
        longitude = point.longitude * toRad
        latitude1 = point1.latitude * toRad
        latitude = point.latitude * toRad
        val dLon = longitude1 - longitude
        val dLambda = latitude1 - latitude
        val dr2 = 2.0 * (1 + (1.0 - Math.cos(dLon)) * Math.cos(latitude1) * Math.cos(latitude)
                - Math.cos(dLambda))
        dr = Math.sqrt(dr2)
        dr_tay = Math.sin(dLambda) - (1 - Math.cos(dLon)) * Math.cos(latitude1) * Math.cos(latitude)
        //if (dr==0)
        //    throw  new NoDirectionException ();
        bearing = (Math.acos(dr_tay / dr) / toRad).toFloat() // 2 roots
        if (dLon < 0) bearing = 360.0.toFloat() - bearing
        println("bearing=" + bearing + " points=" + Arrays.asList(*points))
        return bearing
    }

    fun round(point: LatLng): LatLng {
        val lat = Math.round(point.latitude * eps) / eps
        val lon = Math.round(point.longitude * eps) / eps
        return LatLng(lat, lon)
    }

    class MissMatchDataException : Exception()
    class NoDirectionException : Exception()
}